ouroboros-network/src/Ouroboros/Network/Socket.hs

```diff
@@ -435,67 +444,60 @@ runServerThread
        , Show vNumber
        )
     => NetworkServerTracers ptcl vNumber
-    -> NetworkMutableState
-    -> Socket.Socket
-    -> VersionDataCodec extra CBOR.Term
-    -> (forall vData. extra vData -> vData -> vData -> Accept)
-    -> Versions vNumber extra (OuroborosApplication appType ConnectionId ptcl IO BL.ByteString a b)
+    -> VersionDataCodec vDataT CBOR.Term
+    -> (forall vData . vDataT vData -> vData -> vData -> Accept)
+    -> Versions vNumber vDataT (OuroborosApplication appType ConnectionId ptcl IO BL.ByteString a b)
     -> ErrorPolicies Socket.SockAddr ()
-    -> IO Void
-runServerThread NetworkServerTracers { nstMuxTracer
-                                     , nstHandshakeTracer
-                                     , nstErrorPolicyTracer }
-                NetworkMutableState { nmsConnectionTable
-                                    , nmsPeerStates }
-                sd
-                versionDataCodec
-                acceptVersion
-                versions
-                errorPolicies = do
-    sockAddr <- Socket.getSocketName sd
-    Server.run
-        nstErrorPolicyTracer
-        (fromSocket nmsConnectionTable sd)
-        (acceptException sockAddr)
-        (beginConnection nstMuxTracer nstHandshakeTracer versionDataCodec acceptVersion (acceptConnectionTx sockAddr))
-        -- register producer when application starts, it will be unregistered
-        -- using 'CompleteConnection'
-        (\remoteAddr thread st -> pure $ registerProducer remoteAddr thread
-        st)
-        completeTx mainTx (toLazyTVar nmsPeerStates)
+    -> Connections.ConnectionId -- ^ Includes our address and remote address.
+    -> Socket.Socket            -- ^ Established by the remote peer.
+    -> IO (Connection.Decision IO Remote reject (ConnectionHandle IO))
+incomingConnection NetworkServerTracers { nstMuxTracer
+                                        , nstHandshakeTracer }
+                                        -- , nstErrorPolicyTracer }
+                   versionDataCodec
+                   acceptVersion
+                   versions
+                   -- TODO use these for exception handling.
+                   _errorPolicies
```

Comment:
^ not used error policies - do you plan to integrate them in the follow up PR?

Reply:
Instead of using error policies inside `outgoingConnection` and `incomingConnection` it is better to use them at the call sites of these two functions.

Reply:
Yes this is something I was unsure about. What I gleaned was that the error policies are used by way of `completeApplicationTx`, which would be called whenever a connection ends. This thing has type

```Haskell
completeApplicationTx
  :: forall m addr a.
     ( MonadAsync  m
     , Ord addr
     , Ord (Async m ())
     )
  => ErrorPolicies addr a
  -> CompleteApplication m
       (PeerStates m addr)
       addr
       a
```

but what I'd like to have is

```Haskell
-- Takes the exception which was raised by the connection (or none), and gives the
-- exception which should be thrown.
runErrorPolicies :: ErrorPolicies addr a -> addr -> Maybe SomeException -> (a, Maybe SomeException)
```

Or something similar which can be used to derive

```Haskell
handleWithPolicy :: ErrorPolicies addr a -> addr -> IO () -> IO a
```

Then we can just tack on `handleWithPolicy` to the `runResponder` term.